### PR TITLE
Modernize some documentation with syntax additions and details

### PR DESCRIPTION
**Pull request by NasalMusician**
Is it better to be explicit with ``# Control.Monad.Eff.Effect`` or let the reader figure out that ``# Effect`` can be used after importing ``Control.Monad.Eff``?

### CODE REVIEW

File: language/Types.md [hunk 1]
```diff
@@ -74,7 +74,16 @@ instance showPercentage :: Show Percentage where
...
+Functions in PureScript are like their Javascript counterparts, but always have exactly one argument. Function declarations and lambdas (anonymous functions) with multiple arguments are shorthand for curried lambda abstractions. For example, the following are all equivalent:
+
+```purescript
+add x y = x + y
+add = \x y -> x + y
+add = \x -> \y -> x + y
+add = (+)
+```
+
+Despite this overwhelming convention, uncurried JavaScript functions are can be used through the [purescript-functions](https://pursuit.purescript.org/packages/purescript-functions/) library. For more information see [The Foreign Function Interface](http://www.purescript.org/learn/ffi/).
```
Comment (paf31) [inline language/Types.md]: "overwhelming"? 😄

Comment (NasalMusician) [inline language/Types.md]: Ugh diction. Never works the first time (unlike PureScript!)

Is this better? (also changed the link to point to the relevant section)
> Since uncurried JavaScript functions can be more efficient to run, they are provided indirectly through the [purescript-functions](https://pursuit.purescript.org/packages/purescript-functions/) library. This is unnecessary, however, in most cases and should **not** be used for APIs. For more information see [Functions of Multiple Arguments](http://www.purescript.org/learn/ffi/#functions-of-multiple-arguments).

File: language/Types.md [hunk 2]
```diff
@@ -152,54 +161,88 @@ since the skolemized type variable ``a`` does not unify with ``Int``.
...
 
 ```purescript
-(name :: String, age :: Number | r)
+( name :: String, age :: Number | r )
 ```
 
 ## Type Synonyms
 
-For convenience, it is possible to declare a synonym for a type using the ``type`` keyword. Type synonyms can include type arguments.
+For convenience, it is possible to declare a synonym for a type using the ``type`` keyword. Type synonyms can include type arguments but [cannot be partially applied](https://github.com/purescript/documentation/blob/master/errors/PartiallyAppliedSynonym.md#partiallyappliedsynonym-error). Type synonyms can be built with any other types but [cannot refer to each other in a cycle](https://github.com/purescript/documentation/blob/master/errors/CycleInTypeSynonym.md#cycleintypesynonym-error).
```
Comment (hdgarrood) [inline language/Types.md]: Could you use relative links here please?

File: language/Types.md [hunk 3]
```diff
@@ -152,54 +161,88 @@ since the skolemized type variable ``a`` does not unify with ``Int``.
 
 ## Rows
 
-A row of types represents an unordered collection of named types without duplicates.
+A row of types represents an unordered collection of named types. Duplicate labels are allowed but are distinct.
```
Comment (hdgarrood) [inline language/Types.md]: I think this is maybe a little unclear; possibly we should use the same text as from the release notes here.

File: language/Types.md (hunk 1 above)
Comment (hdgarrood) [inline language/Types.md]: I think this new paragraph is a bit outside the scope of this file; I'm not sure this should be added here.

File: language/Types.md (hunk 3 above)
Comment (NasalMusician) [inline language/Types.md]: This?
> Rows are now unordered collections (of labels and types) with duplicates. However, the collection of types for a specific label within a row is ordered. Conceptually, a row can be thought of as a type-level Map Label (NonEmptyList Type).

What about ...
> A row of types represents an unordered collection of named types, with duplicates. Duplicate labels have their types collected together in order, as if in a ``NonEmptyList``. This means that, conceptually, a row can be thought of as a type-level ``Map Label (NonEmptyList Type)``.

Should we introduce the idea that labels can be strings too? (Just noticing that string labels are non-obvious but used in Halogen/DOM.)

File: language/Types.md (hunk 1 above)
Comment (NasalMusician) [inline language/Types.md]: Well … I think the statement needs *some* explanation for those who do not have functional programming backgrounds (the heresy! 😉 ). So we at least need to explain like, No guys, you really can give multiple arguments to functions!! It just works a bit differently. And at the type level we pretend like everything takes one argument. Maybe a link to a good explanation of currying would be appropriate?

I added the link to the library to encourage readers' natural curiosity. I agree, this file's job is not to explain everything, but I do think the different bits of documentation need to be better linked together to encourage exploration beyond "functions take a single argument". Because what are these types but functions that take more than one argument? Plus, even though they're supported by the compiler, it's all basically just JS/PS code running it! I actually think this library is a prime example of PureScript's flexibility and it would be awesome to encourage learners to be curious about all of this.

Thanks for the feedback, I will keep pondering it.

File: language/Pattern-Matching.md [hunk 4]
```diff
@@ -33,6 +33,7 @@ The following pattern types are supported:
 - Record patterns
 - Named patterns
 - Guards
+- Pattern guards
```
Comment (hdgarrood) [inline language/Pattern-Matching.md]: Pattern guards are not a type of pattern, they're a context in which patterns can be used, so I don't think they should be in this list.

File: language/Types.md [hunk 5]
```diff
@@ -152,54 +176,88 @@ since the skolemized type variable ``a`` does not unify with ``Int``.
...
+
+-- Create an alias for a complex function type
+type Baz = Number -> Number -> Bar Number
+
+-- This function will take two arguments and return a record with double the value
+mkDoubledFoo :: Baz
+mkDoubledFoo foo bar = { foo: 2.0*foo, bar: 2.0*bar }
+
+-- Build on our preious functions to double the values inside any Foo
+-- (Rembmer that Bar Number is the same as Foo)
```
Comment (hdgarrood) [inline language/Types.md]: Typo here?

File: language/Pattern-Matching.md (hunk 4 above)
Comment (hdgarrood) [inline language/Pattern-Matching.md]: I suppose "Guards" shouldn't really be there either, actually, in that case..?

Comment (NasalMusician) [inline language/Pattern-Matching.md]: Is there any place where patterns are allowed but guards are not? View patterns in Haskell even desugar to guards, so I would lean towards saying they are _at least_ intimately related concepts.

Comment (hdgarrood) [inline language/Pattern-Matching.md]: Yes, of course I agree that they are intimately related, and it's worth having discussion of them in this file; I just think it's incorrect to include them in this list.

Comment (NasalMusician) [inline language/Pattern-Matching.md]: Ah, sorry I misunderstood. I personally cannot think of a better way to integrate them, though. Would you like a note afterward like "Guards and pattern guards are also supported"?

I could also hyperlink them, that might be nice …

File: language/Types.md (hunk 5 above)
Comment (NasalMusician) [inline language/Types.md]: Thanks

File: language/Types.md (hunk 1 above)
Comment (NasalMusician) [inline language/Types.md]: Okay ... What about a very short note with two links? Or just revert the changes?

> Functions in PureScript (defined as `input -> output`) are like their JavaScript counterparts, but are curried to always have exactly one argument. Currying works just like [Haskell](https://wiki.haskell.org/Currying).
>
> Uncurried JavaScript functions are also available through the [purescript-functions](https://pursuit.purescript.org/packages/purescript-functions/) library.

The reason I include the link to Haskell is that it explains several features of currying quite well: function returns function, partial application, associativity, pointfree, etc.

Comment (hdgarrood) [inline language/Types.md]: Yes, of course we do need to explain this somewhere; I just think it's important to make sure this particular part of our documentation is focused, to keep it concise and to make sure that people don't have to wade through stuff they already know to get to what they're interested in. If they're here, they probably know what they're looking for (remember that this part of the repo is a language reference, and that is how language references are meant to be used). We have PureScript by Example for people who don't have FP backgrounds or indeed anyone new to PureScript - currying and the purescript-functions library are both covered there.

I do think it would be best to revert these changes - although I'd like to hear from other maintainers too.

File: language/Pattern-Matching.md (hunk 4 above)
Comment (hdgarrood) [inline language/Pattern-Matching.md]: That sounds good to me, yeah. @paf31 / @garyb, thoughts?

**Comment by NasalMusician** [comment]
Sorry for the mess, should I squash it and create a new PR? Or I could categorize/separate out the changes better.
